File: dock_spawn/dock/dock_model.dart
part of dock_spawn;

/**
 * The Dock Model contains the tree hierarchy that represents the state of the 
 * panel placement within the dock manager.
 */ 
class DockModel {
  DockNode rootNode;
  DockNode documentManagerNode;
  
  DockModel() {
  }
}


class DockNode {
  /** The dock container represented by this node */
  IDockContainer container;
  
  
  DockNode parent;
  List<DockNode> children;
  
  DockNode(this.container) {
    children = new List<DockNode>();
  }
  
  void detachFromParent() {
    if (parent != null) {
      parent.removeChild(this);
    }
    parent = null;
  }
  
  void removeChild(DockNode childNode) {
    int index = children.indexOf(childNode);
    if (index >= 0) {
      children.removeRange(index, 1);
    }
  }
  
  void addChild(DockNode childNode) {
    childNode.detachFromParent();
    childNode.parent = this;
    children.add(childNode);
  }

  void addChildBefore(DockNode referenceNode, DockNode childNode) {
    _addChildWithDirection(referenceNode, childNode, true);
  }
  
  void addChildAfter(DockNode referenceNode, DockNode childNode) {
    _addChildWithDirection(referenceNode, childNode, false);
  }
  
  void _addChildWithDirection(DockNode referenceNode, DockNode childNode, bool before) {
    // Detach this node from it's parent first
    childNode.detachFromParent();
    childNode.parent = this;
    
    int referenceIndex = children.indexOf(referenceNode);
    List<DockNode> preList = children.getRange(0, referenceIndex);
    List<DockNode> postList = children.getRange(referenceIndex + 1, children.length - (referenceIndex + 1));
    
    children = new List<DockNode>();
    children.addAll(preList);
    if (before) {
      children.add(childNode);
      children.add(referenceNode);
    } else {
      children.add(referenceNode);
      children.add(childNode);
    }
    children.addAll(postList);
  }
  
  void performLayout() {
    var childContainers = new List<IDockContainer>();
    children.forEach((childNode) => childContainers.add(childNode.container));
    container.performLayout(childContainers);
  }

}

File: js/lib/dock/DockModel.js
/**
 * The Dock Model contains the tree hierarchy that represents the state of the
 * panel placement within the dock manager.
 */
dockspawn.DockModel = function()
{
    this.rootNode = this.documentManagerNode = undefined;
};

dockspawn.DockNode = function(container)
{
    /** The dock container represented by this node */
    this.container = container;
    this.children = [];
}

dockspawn.DockNode.prototype.detachFromParent = function()
{
    if (this.parent)
    {
        this.parent.removeChild(this);
        delete this.parent;
    }
};

dockspawn.DockNode.prototype.removeChild = function(childNode)
{
    var index = this.children.indexOf(childNode);
    if (index >= 0)
        this.children.splice(index, 1);
};

dockspawn.DockNode.prototype.addChild = function(childNode)
{
    childNode.detachFromParent();
    childNode.parent = this;
    this.children.push(childNode);
};

dockspawn.DockNode.prototype.addChildBefore = function(referenceNode, childNode)
{
    this._addChildWithDirection(referenceNode, childNode, true);
};

dockspawn.DockNode.prototype.addChildAfter = function(referenceNode, childNode)
{
    this._addChildWithDirection(referenceNode, childNode, false);
};

dockspawn.DockNode.prototype._addChildWithDirection = function(referenceNode, childNode, before)
{
    // Detach this node from it's parent first
    childNode.detachFromParent();
    childNode.parent = this;

    var referenceIndex = this.children.indexOf(referenceNode);
    var preList = this.children.slice(0, referenceIndex);
    var postList = this.children.slice(referenceIndex + 1, this.children.length);

    this.children = preList.slice(0);
    if (before)
    {
        this.children.push(childNode);
        this.children.push(referenceNode);
    }
    else
    {
        this.children.push(referenceNode);
        this.children.push(childNode);
    }
    Array.prototype.push.apply(this.children, postList);
};

dockspawn.DockNode.prototype.performLayout = function()
{
    var childContainers = this.children.map(function(childNode) { return childNode.container; });
    this.container.performLayout(childContainers);
};

dockspawn.DockNode.prototype.debug_DumpTree = function(indent)
{
    if (indent === undefined)
        indent = 0;

    var message = this.container.name;
    for (var i = 0; i < indent; i++)
        message = "\t" + message;

    var parentType = this.parent === undefined ? "null" : this.parent.container.containerType;
    console.log(">>" + message + " [" + parentType + "]");

    this.children.forEach(function(childNode) { childNode.debug_DumpTree(indent + 1) });
};
